tests/test_main.py

```diff
@@ -0,0 +1,318 @@
+from moodlecli.main import cli
+
+TEST_MOODLE_URL = "http://localhost:8000"
+TEST_MOODLE_TOKEN = "94f9362009170815d912b149bc1e5ffa"
+
+MOODLE_WEBSERVICE_PATH = "/webservice/rest/server.php"
+
+
+def get_matching_helper(request, context):
+    wsfunction = request.qs["wsfunction"][0]
+
+    if wsfunction == moodle.MOODLE_FUNC_GET_ROLE_BY_SHORTNAME:
+        return {'id': 2}
+    elif wsfunction == moodle.MOODLE_FUNC_GET_USERS:
+        # Used to exercise error in course_bulk_setup
+        if request.qs["criteria[0][value]"][0] == 'invalid@gmail.com':
+            return {'exception': 'value'}
+        else:
+            return {'users': [{'id': 2}]}
+    elif wsfunction == moodle.MOODLE_FUNC_GET_SELF_ENROLMENT_METHODS:
+        return [{'id': 2}]
+    else:
+        return []
+
+
+def post_matching_helper(request, context):
+
+    wsfunction = parse.parse_qs(request.body)['wsfunction'][0]
+
+    if wsfunction == moodle.MOODLE_FUNC_DUPLICATE_COURSE:
+        return {'id': 3}
+    elif wsfunction == moodle.MOODLE_FUNC_ENABLE_SELF_ENROLMENT_METHOD:
+        return {'id': 2}
+    elif wsfunction == moodle.MOODLE_FUNC_CREATE_USERS:
+        return {'users': {'id': 5}}
+    elif wsfunction == moodle.MOODLE_FUNC_IMPORT_COURSE:
+        return {'id': 3}
+    elif wsfunction == moodle.MOODLE_FUNC_SET_SELF_ENROLMENT_METHOD_KEY:
+        return {'id': 3}
+    else:
+        return []
+
+
+def test_copy_course(requests_mock):
+    runner = CliRunner()
+    requests_mock.post(f'{TEST_MOODLE_URL}{MOODLE_WEBSERVICE_PATH}',
+                       json=post_matching_helper)
+
+    # copy-course <SOURCE_ID> <COURSE_NAME> <COURSE_SHORT_NAME>
+    # <COURSE_CATEGORY_ID>
+    result = runner.invoke(cli, ['copy-course', '3', 'Algebra3', 'ALG3', '1'],
+                           env={'MOODLE_URL': TEST_MOODLE_URL,
+                                'MOODLE_TOKEN': TEST_MOODLE_TOKEN})
+    assert result.exit_code == 0
+
+
+def test_course_bulk_csv(requests_mock, tmp_path):
+    test_json = {'foo': 'bar'}
+    runner = CliRunner()
+    with runner.isolated_filesystem(temp_dir=tmp_path):
+        requests_mock.get(f'{TEST_MOODLE_URL}{MOODLE_WEBSERVICE_PATH}',
+                          json=test_json)
+        # course-bulk-csv  <OUTPUT_CSV>
+        result = runner.invoke(cli, ['course-bulk-csv', 'output.csv'],
+                               env={'MOODLE_URL': TEST_MOODLE_URL,
+                                    'MOODLE_TOKEN': TEST_MOODLE_TOKEN})
+        assert result.exit_code == 0
+        assert os.stat("output.csv").st_size != 0
+
+
+def test_course_bulk_setup(requests_mock, tmp_path):
+    runner = CliRunner()
+    with runner.isolated_filesystem(temp_dir=tmp_path):
+        requests_mock.get(f'{TEST_MOODLE_URL}{MOODLE_WEBSERVICE_PATH}',
+                          json=get_matching_helper)
+        requests_mock.post(f'{TEST_MOODLE_URL}{MOODLE_WEBSERVICE_PATH}',
+                           json=post_matching_helper)
+
+        f = open('test.csv', 'w')
```

Comment:
Minor: If you wanted to avoid having to remember to `f.close()` in a few of these tests, you can use a context manager here instead:

```python
with open('test.csv, 'w'') as f:
        writer = csv.DictWriter(
                f,
                utils.course_bulk_input_csv_fieldnames()
            )
        writer.writeheader()
        writer.writerows([{utils.CSV_INST_FNAME: 'Tom',
                           utils.CSV_INST_LNAME: 'Michaels',
                           utils.CSV_INST_EMAIL: 'tommichaels@gmail.com',
                           utils.CSV_INST_AUTH: 'manual',
                           utils.CSV_COURSE_NAME: 'Algebra',
                           utils.CSV_COURSE_SHORTNAME: 'a1',
                           utils.CSV_COURSE_CATEGORY: 'MISC'},
                          {utils.CSV_INST_FNAME: 'Christy',
                           utils.CSV_INST_LNAME: 'Yolanda',
                           utils.CSV_INST_EMAIL: 'christyyolanda@gmail.com',
                           utils.CSV_INST_AUTH: 'manual',
                           utils.CSV_COURSE_NAME: 'English',
                           utils.CSV_COURSE_SHORTNAME: 'e1',
                           utils.CSV_COURSE_CATEGORY: 'MISC'}])
```